clis/teliod/src/core_api.rs

```diff
@@ -0,0 +1,343 @@
+}
+
+#[derive(Default, Serialize, Deserialize)]
+struct MeshConfig {
+    public_key: PublicKey,
+    hardware_identifier: String,
+    os: String,
+    os_version: String,
+    device_type: String,
+    traffic_routing_supported: bool,
+}
+
+fn build_exp_backoff() -> Result<ExponentialBackoff, Error> {
+    let backoff_bounds = ExponentialBackoffBounds {
+        initial: Duration::from_secs(1),
+        maximal: Some(Duration::from_secs(300)),
+    };
+    Ok(ExponentialBackoff::new(backoff_bounds)?)
+}
+
+pub async fn init_with_api(
+    auth_token: &str,
+    interface_name: &str,
+) -> Result<DeviceIdentity, Error> {
+    let mut identity_path = dirs::data_local_dir().ok_or(Error::NoDataLocalDir)?;
+    identity_path.push("teliod");
+    if !identity_path.exists() {
+        let _ = create_dir_all(&identity_path);
+    }
+    identity_path.push(&format!("{interface_name}.json"));
+
+    let mut device_identity = DeviceIdentity::new(&identity_path).await;
+
+    if device_identity.machine_identifier.is_none() {
+        device_identity.machine_identifier = Some(
+            match fetch_identifier_with_exp_backoff(
+                auth_token,
+                device_identity.private_key.public(),
+            )
+            .await
+            {
+                Ok(id) => id,
+                Err(e) => {
+                    debug!("Unable to load identifier due to {e}. Registering ...");
+                    register_machine_with_exp_backoff(
+                        &device_identity.hw_identifier,
+                        device_identity.private_key.public(),
+                        auth_token,
+                    )
+                    .await?
+                }
+            },
+        );
+    }
+
+    let mut backoff = build_exp_backoff()?;
+    let mut retries = 0;
+    loop {
+        let status = update_machine(&device_identity, auth_token).await?;
+        match status {
+            StatusCode::OK => break,
+            StatusCode::FORBIDDEN | StatusCode::UNAUTHORIZED | StatusCode::BAD_REQUEST => {
+                // exit daemon
+                return Err(Error::UpdateMachineError(status));
+            }
+            StatusCode::NOT_FOUND => {
+                debug!("Unable to update. Registering machine ...");
+                // Retry machine update after registering. To make sure everything was successful
+                // If registering fails. Close the daemon
+                device_identity.machine_identifier = Some(
+                    register_machine_with_exp_backoff(
+                        &device_identity.hw_identifier,
+                        device_identity.private_key.public(),
+                        auth_token,
+                    )
+                    .await?,
+                );
+            }
+            _ => {
+                // Retry with exp back-off
+                let _ =
+                    retry_after_backoff_delay(&mut backoff, &mut retries, "update machine").await?;
```

Comment:
Tbh Telio's exponential backoff might not be the best tool for this job - it was intended to be used with Telio's state machines (it could be done in this way, but probably not worth it) - I think using https://docs.rs/backoff/0.4.0/backoff/ you could greatly simplify this code - probably with `[with_max_elapsed_time](https://docs.rs/backoff/0.4.0/backoff/exponential/struct.ExponentialBackoffBuilder.html#method.with_max_elapsed_time)` you wouldn't even need `retries`

Reply:
Changed to `backoff` crate